pkg/action/cl001/ac008/executor.go

```diff
@@ -2,8 +2,145 @@ package ac008
+
+	masterNodeLabelSelector = "kubernetes.io/role=master"
+	workerNodeLabelSelector = "kubernetes.io/role=worker"
 )
 
 func (e *Executor) execute(ctx context.Context) error {
+	var err error
+
+	var cpClients k8sclient.Interface
+	{
+		c := client.ControlPlaneConfig{
+			Logger: e.logger,
+
+			KubeConfig: env.KubeConfig(),
+		}
+
+		cpClients, err = client.NewControlPlane(c)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	var tcClients k8sclient.Interface
+	{
+		c := client.TenantClusterConfig{
+			ControlPlane: cpClients,
+			Logger:       e.logger,
+
+			Scope: e.scope,
+		}
+
+		tcClients, err = client.NewTenantCluster(c)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	err = e.checkTLSCerts(ctx, tcClients.K8sClient())
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	err = e.checkKiamPods(ctx, tcClients.K8sClient())
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+// checkTLSCerts Ensures that kiam  tls certs are created.
+var kiamTlSCertSecretNames = []string{"kiam-agent-tls", "kiam-ca-tls", "kiam-server-tls"}
+
+func (e *Executor) checkTLSCerts(ctx context.Context, tcClient kubernetes.Interface) error {
+	for _, secret := range kiamTlSCertSecretNames {
+		_, err := tcClient.CoreV1().Secrets(kubeSystemNamespace).Get(ctx, secret, metav1.GetOptions{})
+
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	return nil
+}
+
+// checkKiamPods ensures kiam-agent and kiam-server pods are alive and running
+func (e *Executor) checkKiamPods(ctx context.Context, tcClient kubernetes.Interface) error {
+	// count expected kiam-server and kiam-agent pods
+	var expectedKiamServerPodCount, expectedKiamAgentPodCount int
+	{
+		masterNodes, err := tcClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{LabelSelector: masterNodeLabelSelector})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		expectedKiamServerPodCount = len(masterNodes.Items)
+
+		workerNodes, err := tcClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{LabelSelector: workerNodeLabelSelector})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		expectedKiamAgentPodCount = len(workerNodes.Items)
+	}
+
+	// kiam server
+	{
+		kiamServerPods, err := tcClient.CoreV1().Pods(kubeSystemNamespace).List(ctx, metav1.ListOptions{LabelSelector: kiamServerLabelSelector})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if len(kiamServerPods.Items) != expectedKiamServerPodCount {
+			executionFailedError.Desc = fmt.Sprintf("Execution failed: wrong kiam-server pod count, expected %d but got %d", expectedKiamServerPodCount, len(kiamServerPods.Items))
+			return microerror.Mask(executionFailedError)
+		}
+
+		for _, kiamServerPod := range kiamServerPods.Items {
+			if kiamServerPod.Status.Phase != apiv1.PodRunning {
+				executionFailedError.Desc = fmt.Sprintf("Execution failed: pod %s in namespace %s is not running.", kiamServerPod.Name, kiamServerPod.Namespace)
```

Comment:
We should rather use `microerror.Maskf(executionFailedError, "...")`. The error is a pointer and it global to the package. When you change it here you change it forever. So far the description was rather meant to be static and declared once.

Reply:
ack, saw it somewhere else in awscnf so that's why copied it :p

Reply:
I hope I did not write that code. 🙈

Reply:
sorted